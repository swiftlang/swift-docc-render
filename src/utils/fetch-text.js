/**
 * This source file is part of the Swift.org open source project
 *
 * Copyright (c) 2021 Apple Inc. and the Swift project authors
 * Licensed under Apache License v2.0 with Runtime Library Exception
 *
 * See https://swift.org/LICENSE.txt for license information
 * See https://swift.org/CONTRIBUTORS.txt for Swift project authors
*/

import { resolveAbsoluteUrl } from 'docc-render/utils/url-helper';

/**
 * Fetch the contents of a file as text.
 * @param {string} filepath The file path.
 * @param {RequestInit?} options Optional request settings.
 * @returns {Promise<string>} The text contents of the file.
 */
export default async function fetchText(filepath, options) {
  const url = resolveAbsoluteUrl(filepath);
  return fetch(url, options)
    .then(r => r.text());
}
