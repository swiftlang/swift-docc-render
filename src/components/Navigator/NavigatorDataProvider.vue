<!--
  This source file is part of the Swift.org open source project

  Copyright (c) 2022-2024 Apple Inc. and the Swift project authors
  Licensed under Apache License v2.0 with Runtime Library Exception

  See https://swift.org/LICENSE.txt for license information
  See https://swift.org/CONTRIBUTORS.txt for Swift project authors
-->

<script>
import { fetchIndexPathsData } from 'docc-render/utils/data';
import { flattenNestedData } from 'docc-render/utils/navigatorData';
import AppStore from 'docc-render/stores/AppStore';
import Language from 'docc-render/constants/Language';

/**
 * Fetches the sidebar navigator data and provides it via a scoped slot,
 * extracting for current language
 */
export default {
  name: 'NavigatorDataProvider',
  props: {
    interfaceLanguage: {
      type: String,
      default: Language.swift.key.url,
    },
    /**
     * The technology we need to fetch data for.
     * Important - We are passing just the URL, as the technology object changes
     * between page navigations, resulting in excess re-calculations on each page change.
     */
    technologyUrl: {
      type: String,
      default: '',
    },
    apiChangesVersion: {
      type: String,
      default: '',
    },
  },
  data() {
    return {
      isFetching: false,
      errorFetching: false,
      isFetchingAPIChanges: false,
      navigationIndex: {
        [Language.swift.key.url]: [],
      },
      navigationReferences: {},
      diffs: null,
    };
  },
  computed: {
    /**
     * Recomputes the list of flat children.
     * @return NavigatorFlatItem[]
     */
    flatChildren: ({
      technologyWithChildren = {},
    }) => (
      flattenNestedData(
        technologyWithChildren.children || [], null, 0, technologyWithChildren.beta,
      )
    ),
    technologyPath: ({ technologyUrl }) => {
      // regex should match only the first section, no slash - `/documentation/:technology`
      const matches = /(\/documentation\/(?:[^/]+))\/?/.exec(technologyUrl);
      return matches ? matches[1] : '';
    },
    /**
     * Extracts the technology data, for the currently chosen language
     * @return {Object}
     */
    technologyWithChildren({ navigationIndex, interfaceLanguage, technologyPath }) {
      // get the technologies for the current language
      let currentLangTechnologies = navigationIndex[interfaceLanguage] || [];
      // if no such items, we use the default swift one
      if (!currentLangTechnologies.length) {
        currentLangTechnologies = navigationIndex[Language.swift.key.url] || [];
      }
      // find the current technology
      return currentLangTechnologies.find(t => (
        technologyPath.toLowerCase() === t.path.toLowerCase()
      ));
    },
  },
  methods: {
    async fetchIndexData() {
      try {
        this.isFetching = true;
        const {
          includedArchiveIdentifiers = [],
          interfaceLanguages,
          references,
        } = await fetchIndexPathsData(
          { slug: this.$route.params.locale || '' },
        );
        this.navigationIndex = Object.freeze(interfaceLanguages);
        this.navigationReferences = Object.freeze(references);
        AppStore.setIncludedArchiveIdentifiers(includedArchiveIdentifiers);
      } catch (e) {
        this.errorFetching = true;
      } finally {
        this.isFetching = false;
      }
    },
  },
  watch: {
    '$route.params.locale': {
      handler: 'fetchIndexData',
      immediate: true,
    },
  },
  render() {
    return this.$scopedSlots.default({
      technology: this.technologyWithChildren,
      isFetching: this.isFetching,
      errorFetching: this.errorFetching,
      isFetchingAPIChanges: this.isFetchingAPIChanges,
      apiChanges: this.diffs,
      flatChildren: this.flatChildren,
      references: this.navigationReferences,
    });
  },
};
</script>
