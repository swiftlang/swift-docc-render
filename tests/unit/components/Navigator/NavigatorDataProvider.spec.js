/**
 * This source file is part of the Swift.org open source project
 *
 * Copyright (c) 2022 Apple Inc. and the Swift project authors
 * Licensed under Apache License v2.0 with Runtime Library Exception
 *
 * See https://swift.org/LICENSE.txt for license information
 * See https://swift.org/CONTRIBUTORS.txt for Swift project authors
*/

import NavigatorDataProvider from '@/components/Navigator/NavigatorDataProvider.vue';
import { shallowMount } from '@vue/test-utils';
import AppStore from 'docc-render/stores/AppStore';
import Language from 'docc-render/constants/Language';
import { TopicTypes } from '@/constants/TopicTypes';
import { fetchIndexPathsData } from '@/utils/data';
import { INDEX_ROOT_KEY } from '@/constants/sidebar';
import { flushPromises } from '../../../../test-utils';

jest.mock('docc-render/utils/data');

const technologyUrl = '/documentation/foo';

const extendedTechnologies = {
  path: '/documentation/foo',
  children: [
    {
      id: 'Group Marker',
      title: 'Group Marker',
      type: TopicTypes.groupMarker,
    },
    {
      id: 'Child0',
      title: 'Child0',
      path: '/documentation/foo/child0',
      type: 'article',
      children: [
        {
          id: 'Group Marker, Child 0',
          title: 'Group Marker, Child 0',
          type: TopicTypes.groupMarker,
        },
        {
          id: 'Child0_GrandChild0',
          title: 'Child0_GrandChild0',
          path: '/documentation/foo/child0/grandchild0',
          type: 'tutorial',
        },
        {
          id: 'Child0_GrandChild1',
          title: 'Child0_GrandChild1',
          path: '/documentation/foo/child0/grandchild1',
          type: 'tutorial',
          children: [{
            id: 'Child0_GrandChild0_GreatGrandChild0',
            title: 'Child0_GrandChild0_GreatGrandChild0',
            path: '/documentation/foo/child0/grandchild0/greatgrandchild0',
            type: 'tutorial',
          }],
        },
        {
          id: 'Child0_GrandChild2',
          title: 'Child0_GrandChild2',
          path: '/documentation/foo/child0/grandchild2',
          type: 'tutorial',
        },
      ],
    },
    {
      id: 'Child1',
      title: 'Child1',
      path: '/documentation/foo/child1/',
      type: 'tutorial',
      children: [{
        id: 'Child1_GrandChild0',
        title: 'Child1_GrandChild0',
        path: '/documentation/foo/child1/grandchild0',
        type: 'method',
      }],
    },
  ],
};

const flatChildren = [
  {
    uid: -196255993,
    parent: '<root>',
    index: 0,
    siblingsCount: 3,
    depth: 0,
    childUIDs: [],
  },
  {
    uid: -196255992,
    parent: '<root>',
    index: 1,
    siblingsCount: 3,
    depth: 0,
    childUIDs: [],
  },
  {
    uid: -196255991,
    parent: '<root>',
    index: 2,
    siblingsCount: 3,
    depth: 0,
    childUIDs: [],
  },
];

const swiftIndexOne = {
  id: 'foo',
  path: technologyUrl,
  children: [1, 2, 3],
};
const swiftIndexTwo = {
  id: 'bar',
  path: '/bar',
  children: [1],
};
const objectiveCIndexOne = {
  id: 'foo-objc',
  path: technologyUrl,
  children: [1],
};

const references = {
  foo: { bar: 'bar' },
};

const includedArchiveIdentifiers = [
  'foo',
  'bar',
];

const response = {
  includedArchiveIdentifiers,
  interfaceLanguages: {
    [Language.swift.key.url]: [
      swiftIndexOne,
      swiftIndexTwo,
    ],
    [Language.objectiveC.key.url]: [
      objectiveCIndexOne,
    ],
  },
  references,
};

fetchIndexPathsData.mockResolvedValue(response);

const defaultProps = {
  technologyUrl,
};

let props = {};

const createWrapper = ({ propsData, ...others } = {}) => shallowMount(NavigatorDataProvider, {
  propsData: {
    ...defaultProps,
    ...propsData,
  },
  scopedSlots: {
    default: (p) => {
      props = p;
      return 'Text';
    },
  },
  mocks: {
    $route: {
      params: {
        locale: 'en-US',
      },
    },
  },
  ...others,
});

describe('NavigatorDataProvider', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  afterEach(() => {
    AppStore.reset();
  });

  it('fetches data when mounting NavigatorDataProvider', async () => {
    expect(fetchIndexPathsData).toHaveBeenCalledTimes(0);
    createWrapper();
    expect(fetchIndexPathsData).toHaveBeenCalledTimes(1);
    expect(props).toHaveProperty('isFetching', true);
    await flushPromises();
    expect(props).toEqual({
      apiChanges: null,
      isFetchingAPIChanges: false,
      isFetching: false,
      technology: swiftIndexOne,
      errorFetching: false,
      flatChildren,
      references,
    });
  });

  it('fetches data, even if being passed a none-root technology url', async () => {
    expect(fetchIndexPathsData).toHaveBeenCalledTimes(0);
    createWrapper({
      propsData: {
        technologyUrl: `${technologyUrl}/bar/baz`,
      },
    });
    expect(fetchIndexPathsData).toHaveBeenCalledTimes(1);
    expect(props).toHaveProperty('isFetching', true);
    await flushPromises();
    expect(props).toEqual({
      apiChanges: null,
      isFetchingAPIChanges: false,
      isFetching: false,
      technology: swiftIndexOne,
      errorFetching: false,
      flatChildren,
      references,
    });
  });

  it('sets errorFetching to true, when request errored', async () => {
    expect(fetchIndexPathsData).toHaveBeenCalledTimes(0);
    fetchIndexPathsData.mockRejectedValueOnce('Error');
    createWrapper();
    expect(fetchIndexPathsData).toHaveBeenCalledTimes(1);
    expect(props).toHaveProperty('isFetching', true);
    await flushPromises();
    expect(props).toEqual({
      apiChanges: null,
      isFetchingAPIChanges: false,
      isFetching: false,
      technology: undefined,
      errorFetching: true,
      flatChildren: [],
      references: {},
    });
  });

  it('returns objc data', async () => {
    expect(fetchIndexPathsData).toHaveBeenCalledTimes(0);
    createWrapper({
      propsData: {
        interfaceLanguage: Language.objectiveC.key.url,
      },
    });
    expect(fetchIndexPathsData).toHaveBeenCalledTimes(1);
    expect(props).toHaveProperty('isFetching', true);
    await flushPromises();
    expect(props).toEqual({
      apiChanges: null,
      isFetchingAPIChanges: false,
      errorFetching: false,
      isFetching: false,
      technology: objectiveCIndexOne,
      flatChildren: [{
        uid: -196255993,
        parent: '<root>',
        index: 0,
        siblingsCount: 1,
        depth: 0,
        childUIDs: [],
      }],
      references,
    });
  });

  it('falls back to swift items, if no objc items', async () => {
    expect(fetchIndexPathsData).toHaveBeenCalledTimes(0);
    fetchIndexPathsData.mockResolvedValueOnce({
      interfaceLanguages: {
        [Language.swift.key.url]: response.interfaceLanguages[Language.swift.key.url],
      },
      references,
    });
    createWrapper({
      propsData: {
        interfaceLanguage: Language.objectiveC.key.url,
      },
    });
    expect(fetchIndexPathsData).toHaveBeenCalledTimes(1);
    expect(props).toHaveProperty('isFetching', true);
    await flushPromises();
    expect(props).toEqual({
      apiChanges: null,
      isFetchingAPIChanges: false,
      errorFetching: false,
      isFetching: false,
      technology: swiftIndexOne,
      flatChildren,
      references,
    });
  });

  it('returns undefined technology, if none matches', async () => {
    createWrapper({
      propsData: {
        technologyUrl: '/documentation/bar',
      },
    });
    await flushPromises();
    expect(props).toEqual({
      apiChanges: null,
      isFetchingAPIChanges: false,
      errorFetching: false,
      isFetching: false,
      technology: undefined,
      flatChildren: [],
      references,
    });
  });

  it('returns undefined technology when index response is empty', async () => {
    const emptyResponse = { interfaceLanguages: {} };
    fetchIndexPathsData.mockResolvedValue(emptyResponse);
    createWrapper();
    await flushPromises();
    expect(props).toEqual({
      apiChanges: null,
      isFetchingAPIChanges: false,
      isFetching: false,
      technology: undefined,
      errorFetching: false,
      flatChildren: [],
      references: undefined,
    });
  });

  it('counts the amount of deprecated items a groupMarker has', async () => {
    const technologyClone = JSON.parse(JSON.stringify(extendedTechnologies));
    technologyClone.children[1].deprecated = true;
    technologyClone.children[2].deprecated = true;
    technologyClone.children[1].children[1].deprecated = true;
    fetchIndexPathsData.mockResolvedValue({
      interfaceLanguages: {
        [Language.swift.key.url]: [
          technologyClone,
        ],
      },
    });
    createWrapper();
    await flushPromises();
    expect(props.flatChildren[0]).toHaveProperty('deprecatedChildrenCount', 2);
    expect(props.flatChildren).toMatchSnapshot();
  });

  it('removes the `beta` flag from children, if the parent is a `beta`', () => {
    const technologyClone = JSON.parse(JSON.stringify(extendedTechnologies));
    technologyClone.beta = true;
    technologyClone.children[1].beta = true;
    technologyClone.children[1].children[0].beta = true;
    fetchIndexPathsData.mockResolvedValue({
      interfaceLanguages: {
        [Language.swift.key.url]: [
          technologyClone,
        ],
      },
    });
    createWrapper();
    expect(props).toMatchSnapshot();
  });

  it('removes the `beta` flag from children, if the parent is a `beta`', () => {
    const technologyClone = JSON.parse(JSON.stringify(extendedTechnologies));
    technologyClone.children[1].beta = true;
    technologyClone.children[1].children[1].beta = true;
    // case where the direct parent is NOT `Beta`, but an ancestor is
    technologyClone.children[1].children[2].children[0].beta = true;
    // set an end node as beta
    technologyClone.children[2].children[0].beta = true;
    fetchIndexPathsData.mockResolvedValue({
      interfaceLanguages: {
        [Language.swift.key.url]: [
          technologyClone,
        ],
      },
    });
    createWrapper();
    expect(props).toMatchSnapshot();
  });

  it('flattens deeply nested children and provides them to the NavigatorCard', async () => {
    fetchIndexPathsData.mockResolvedValue({
      interfaceLanguages: {
        [Language.swift.key.url]: [{
          path: '/documentation/foo',
          children: [
            {
              title: 'Group Marker',
              type: TopicTypes.groupMarker,
            },
            {
              title: 'Child0',
              path: '/foo/child0',
              type: 'article',
              children: [
                {
                  title: 'Group Marker, Child 0',
                  type: TopicTypes.groupMarker,
                },
                {
                  title: 'Child0_GrandChild0',
                  path: '/foo/child0/grandchild0',
                  type: 'tutorial',
                },
                {
                  title: 'Child0_GrandChild1',
                  path: '/foo/child0/grandchild1',
                  type: 'tutorial',
                  children: [{
                    title: 'Child0_GrandChild0_GreatGrandChild0',
                    path: '/foo/child0/grandchild0/greatgrandchild0',
                    type: 'tutorial',
                  }],
                },
                {
                  title: 'Child0_GrandChild2',
                  path: '/foo/child0/grandchild2',
                  type: 'tutorial',
                },
              ],
            },
            {
              title: 'Child1',
              path: '/foo/child1/',
              type: 'tutorial',
              children: [{
                title: 'Child1_GrandChild0',
                path: '/foo/child1/grandchild0',
                type: 'method',
              }],
            },
          ],
        }],
      },
    });
    createWrapper();
    await flushPromises();
    expect(props.flatChildren).toEqual([
      {
        childUIDs: [
          551503844,
          -97593391,
        ],
        deprecatedChildrenCount: 0,
        depth: 0,
        index: 0,
        parent: INDEX_ROOT_KEY,
        siblingsCount: 3,
        title: 'Group Marker',
        type: 'groupMarker',
        uid: -196255993,
      },
      {
        childUIDs: [
          -361407047,
          1438225895,
          1439149417,
          1440072939,
        ],
        depth: 0,
        groupMarkerUID: -196255993,
        index: 1,
        parent: INDEX_ROOT_KEY,
        path: '/foo/child0',
        siblingsCount: 3,
        title: 'Child0',
        type: 'article',
        uid: 551503844,
      },
      {
        childUIDs: [
          1438225895,
          1439149417,
          1440072939,
        ],
        deprecatedChildrenCount: 0,
        depth: 1,
        index: 0,
        parent: 551503844,
        siblingsCount: 4,
        title: 'Group Marker, Child 0',
        type: 'groupMarker',
        uid: -361407047,
      },
      {
        childUIDs: [],
        depth: 1,
        groupMarkerUID: -361407047,
        index: 1,
        parent: 551503844,
        path: '/foo/child0/grandchild0',
        siblingsCount: 4,
        title: 'Child0_GrandChild0',
        type: 'tutorial',
        uid: 1438225895,
      },
      {
        childUIDs: [
          305326087,
        ],
        depth: 1,
        groupMarkerUID: -361407047,
        index: 2,
        parent: 551503844,
        path: '/foo/child0/grandchild1',
        siblingsCount: 4,
        title: 'Child0_GrandChild1',
        type: 'tutorial',
        uid: 1439149417,
      },
      {
        childUIDs: [],
        depth: 2,
        index: 0,
        parent: 1439149417,
        path: '/foo/child0/grandchild0/greatgrandchild0',
        siblingsCount: 1,
        title: 'Child0_GrandChild0_GreatGrandChild0',
        type: 'tutorial',
        uid: 305326087,
      },
      {
        childUIDs: [],
        depth: 1,
        groupMarkerUID: -361407047,
        index: 3,
        parent: 551503844,
        path: '/foo/child0/grandchild2',
        siblingsCount: 4,
        title: 'Child0_GrandChild2',
        type: 'tutorial',
        uid: 1440072939,
      },
      {
        childUIDs: [
          -827353283,
        ],
        depth: 0,
        groupMarkerUID: -196255993,
        index: 2,
        parent: INDEX_ROOT_KEY,
        path: '/foo/child1/',
        siblingsCount: 3,
        title: 'Child1',
        type: 'tutorial',
        uid: -97593391,
      },
      {
        childUIDs: [],
        depth: 1,
        index: 0,
        parent: -97593391,
        path: '/foo/child1/grandchild0',
        siblingsCount: 1,
        title: 'Child1_GrandChild0',
        type: 'method',
        uid: -827353283,
      },
    ]);
  });

  it('sets `includedArchiveIdentifiers` state in the app store', async () => {
    expect(AppStore.state.includedArchiveIdentifiers).toEqual([]);
    fetchIndexPathsData.mockResolvedValue(response);
    createWrapper();
    await flushPromises();
    expect(AppStore.state.includedArchiveIdentifiers).toEqual(includedArchiveIdentifiers);
  });
});
