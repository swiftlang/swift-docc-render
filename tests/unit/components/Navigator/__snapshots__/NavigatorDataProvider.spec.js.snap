// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavigatorDataProvider counts the amount of deprecated items a groupMarker has 1`] = `
Array [
  Object {
    "childUIDs": Array [
      365745369,
      -1561138820,
    ],
    "deprecatedChildrenCount": 2,
    "depth": 0,
    "id": "Group Marker",
    "index": 0,
    "parent": <root>,
    "siblingsCount": 3,
    "title": "Group Marker",
    "type": "groupMarker",
    "uid": -196255993,
  },
  Object {
    "childUIDs": Array [
      -1893192488,
      -2046366661,
      -2045443139,
      -2044519617,
    ],
    "deprecated": true,
    "depth": 0,
    "groupMarkerUID": -196255993,
    "id": "Child0",
    "index": 1,
    "parent": <root>,
    "path": "/documentation/foo/child0",
    "siblingsCount": 3,
    "title": "Child0",
    "type": "article",
    "uid": 365745369,
  },
  Object {
    "childUIDs": Array [
      -2046366661,
      -2045443139,
      -2044519617,
    ],
    "deprecatedChildrenCount": 1,
    "depth": 1,
    "id": "Group Marker, Child 0",
    "index": 0,
    "parent": 365745369,
    "siblingsCount": 4,
    "title": "Group Marker, Child 0",
    "type": "groupMarker",
    "uid": -1893192488,
  },
  Object {
    "childUIDs": Array [],
    "deprecated": true,
    "depth": 1,
    "groupMarkerUID": -1893192488,
    "id": "Child0_GrandChild0",
    "index": 1,
    "parent": 365745369,
    "path": "/documentation/foo/child0/grandchild0",
    "siblingsCount": 4,
    "title": "Child0_GrandChild0",
    "type": "tutorial",
    "uid": -2046366661,
  },
  Object {
    "childUIDs": Array [
      366459535,
    ],
    "depth": 1,
    "groupMarkerUID": -1893192488,
    "id": "Child0_GrandChild1",
    "index": 2,
    "parent": 365745369,
    "path": "/documentation/foo/child0/grandchild1",
    "siblingsCount": 4,
    "title": "Child0_GrandChild1",
    "type": "tutorial",
    "uid": -2045443139,
  },
  Object {
    "childUIDs": Array [],
    "depth": 2,
    "id": "Child0_GrandChild0_GreatGrandChild0",
    "index": 0,
    "parent": -2045443139,
    "path": "/documentation/foo/child0/grandchild0/greatgrandchild0",
    "siblingsCount": 1,
    "title": "Child0_GrandChild0_GreatGrandChild0",
    "type": "tutorial",
    "uid": 366459535,
  },
  Object {
    "childUIDs": Array [],
    "depth": 1,
    "groupMarkerUID": -1893192488,
    "id": "Child0_GrandChild2",
    "index": 3,
    "parent": 365745369,
    "path": "/documentation/foo/child0/grandchild2",
    "siblingsCount": 4,
    "title": "Child0_GrandChild2",
    "type": "tutorial",
    "uid": -2044519617,
  },
  Object {
    "childUIDs": Array [
      1069111095,
    ],
    "deprecated": true,
    "depth": 0,
    "groupMarkerUID": -196255993,
    "id": "Child1",
    "index": 2,
    "parent": <root>,
    "path": "/documentation/foo/child1/",
    "siblingsCount": 3,
    "title": "Child1",
    "type": "tutorial",
    "uid": -1561138820,
  },
  Object {
    "childUIDs": Array [],
    "depth": 1,
    "id": "Child1_GrandChild0",
    "index": 0,
    "parent": -1561138820,
    "path": "/documentation/foo/child1/grandchild0",
    "siblingsCount": 1,
    "title": "Child1_GrandChild0",
    "type": "method",
    "uid": 1069111095,
  },
]
`;

exports[`NavigatorDataProvider removes the \`beta\` flag from children, if the parent is a \`beta\` 1`] = `
Object {
  "apiChanges": null,
  "errorFetching": false,
  "flatChildren": Array [
    Object {
      "childUIDs": Array [
        365745369,
        -1561138820,
      ],
      "deprecatedChildrenCount": 0,
      "depth": 0,
      "id": "Group Marker",
      "index": 0,
      "parent": <root>,
      "siblingsCount": 3,
      "title": "Group Marker",
      "type": "groupMarker",
      "uid": -196255993,
    },
    Object {
      "beta": false,
      "childUIDs": Array [
        -1893192488,
        -2046366661,
        -2045443139,
        -2044519617,
      ],
      "depth": 0,
      "groupMarkerUID": -196255993,
      "id": "Child0",
      "index": 1,
      "parent": <root>,
      "path": "/documentation/foo/child0",
      "siblingsCount": 3,
      "title": "Child0",
      "type": "article",
      "uid": 365745369,
    },
    Object {
      "beta": false,
      "childUIDs": Array [
        -2046366661,
        -2045443139,
        -2044519617,
      ],
      "deprecatedChildrenCount": 0,
      "depth": 1,
      "id": "Group Marker, Child 0",
      "index": 0,
      "parent": 365745369,
      "siblingsCount": 4,
      "title": "Group Marker, Child 0",
      "type": "groupMarker",
      "uid": -1893192488,
    },
    Object {
      "childUIDs": Array [],
      "depth": 1,
      "groupMarkerUID": -1893192488,
      "id": "Child0_GrandChild0",
      "index": 1,
      "parent": 365745369,
      "path": "/documentation/foo/child0/grandchild0",
      "siblingsCount": 4,
      "title": "Child0_GrandChild0",
      "type": "tutorial",
      "uid": -2046366661,
    },
    Object {
      "childUIDs": Array [
        366459535,
      ],
      "depth": 1,
      "groupMarkerUID": -1893192488,
      "id": "Child0_GrandChild1",
      "index": 2,
      "parent": 365745369,
      "path": "/documentation/foo/child0/grandchild1",
      "siblingsCount": 4,
      "title": "Child0_GrandChild1",
      "type": "tutorial",
      "uid": -2045443139,
    },
    Object {
      "childUIDs": Array [],
      "depth": 2,
      "id": "Child0_GrandChild0_GreatGrandChild0",
      "index": 0,
      "parent": -2045443139,
      "path": "/documentation/foo/child0/grandchild0/greatgrandchild0",
      "siblingsCount": 1,
      "title": "Child0_GrandChild0_GreatGrandChild0",
      "type": "tutorial",
      "uid": 366459535,
    },
    Object {
      "childUIDs": Array [],
      "depth": 1,
      "groupMarkerUID": -1893192488,
      "id": "Child0_GrandChild2",
      "index": 3,
      "parent": 365745369,
      "path": "/documentation/foo/child0/grandchild2",
      "siblingsCount": 4,
      "title": "Child0_GrandChild2",
      "type": "tutorial",
      "uid": -2044519617,
    },
    Object {
      "childUIDs": Array [
        1069111095,
      ],
      "depth": 0,
      "groupMarkerUID": -196255993,
      "id": "Child1",
      "index": 2,
      "parent": <root>,
      "path": "/documentation/foo/child1/",
      "siblingsCount": 3,
      "title": "Child1",
      "type": "tutorial",
      "uid": -1561138820,
    },
    Object {
      "childUIDs": Array [],
      "depth": 1,
      "id": "Child1_GrandChild0",
      "index": 0,
      "parent": -1561138820,
      "path": "/documentation/foo/child1/grandchild0",
      "siblingsCount": 1,
      "title": "Child1_GrandChild0",
      "type": "method",
      "uid": 1069111095,
    },
  ],
  "isFetching": false,
  "isFetchingAPIChanges": false,
  "references": undefined,
  "technology": Object {
    "beta": true,
    "children": Array [
      Object {
        "id": "Group Marker",
        "title": "Group Marker",
        "type": "groupMarker",
      },
      Object {
        "beta": true,
        "children": Array [
          Object {
            "beta": true,
            "id": "Group Marker, Child 0",
            "title": "Group Marker, Child 0",
            "type": "groupMarker",
          },
          Object {
            "id": "Child0_GrandChild0",
            "path": "/documentation/foo/child0/grandchild0",
            "title": "Child0_GrandChild0",
            "type": "tutorial",
          },
          Object {
            "children": Array [
              Object {
                "id": "Child0_GrandChild0_GreatGrandChild0",
                "path": "/documentation/foo/child0/grandchild0/greatgrandchild0",
                "title": "Child0_GrandChild0_GreatGrandChild0",
                "type": "tutorial",
              },
            ],
            "id": "Child0_GrandChild1",
            "path": "/documentation/foo/child0/grandchild1",
            "title": "Child0_GrandChild1",
            "type": "tutorial",
          },
          Object {
            "id": "Child0_GrandChild2",
            "path": "/documentation/foo/child0/grandchild2",
            "title": "Child0_GrandChild2",
            "type": "tutorial",
          },
        ],
        "id": "Child0",
        "path": "/documentation/foo/child0",
        "title": "Child0",
        "type": "article",
      },
      Object {
        "children": Array [
          Object {
            "id": "Child1_GrandChild0",
            "path": "/documentation/foo/child1/grandchild0",
            "title": "Child1_GrandChild0",
            "type": "method",
          },
        ],
        "id": "Child1",
        "path": "/documentation/foo/child1/",
        "title": "Child1",
        "type": "tutorial",
      },
    ],
    "path": "/documentation/foo",
  },
}
`;

exports[`NavigatorDataProvider removes the \`beta\` flag from children, if the parent is a \`beta\` 2`] = `
Object {
  "apiChanges": null,
  "errorFetching": false,
  "flatChildren": Array [
    Object {
      "childUIDs": Array [
        365745369,
        -1561138820,
      ],
      "deprecatedChildrenCount": 0,
      "depth": 0,
      "id": "Group Marker",
      "index": 0,
      "parent": <root>,
      "siblingsCount": 3,
      "title": "Group Marker",
      "type": "groupMarker",
      "uid": -196255993,
    },
    Object {
      "beta": true,
      "childUIDs": Array [
        -1893192488,
        -2046366661,
        -2045443139,
        -2044519617,
      ],
      "depth": 0,
      "groupMarkerUID": -196255993,
      "id": "Child0",
      "index": 1,
      "parent": <root>,
      "path": "/documentation/foo/child0",
      "siblingsCount": 3,
      "title": "Child0",
      "type": "article",
      "uid": 365745369,
    },
    Object {
      "childUIDs": Array [
        -2046366661,
        -2045443139,
        -2044519617,
      ],
      "deprecatedChildrenCount": 0,
      "depth": 1,
      "id": "Group Marker, Child 0",
      "index": 0,
      "parent": 365745369,
      "siblingsCount": 4,
      "title": "Group Marker, Child 0",
      "type": "groupMarker",
      "uid": -1893192488,
    },
    Object {
      "beta": false,
      "childUIDs": Array [],
      "depth": 1,
      "groupMarkerUID": -1893192488,
      "id": "Child0_GrandChild0",
      "index": 1,
      "parent": 365745369,
      "path": "/documentation/foo/child0/grandchild0",
      "siblingsCount": 4,
      "title": "Child0_GrandChild0",
      "type": "tutorial",
      "uid": -2046366661,
    },
    Object {
      "childUIDs": Array [
        366459535,
      ],
      "depth": 1,
      "groupMarkerUID": -1893192488,
      "id": "Child0_GrandChild1",
      "index": 2,
      "parent": 365745369,
      "path": "/documentation/foo/child0/grandchild1",
      "siblingsCount": 4,
      "title": "Child0_GrandChild1",
      "type": "tutorial",
      "uid": -2045443139,
    },
    Object {
      "beta": false,
      "childUIDs": Array [],
      "depth": 2,
      "id": "Child0_GrandChild0_GreatGrandChild0",
      "index": 0,
      "parent": -2045443139,
      "path": "/documentation/foo/child0/grandchild0/greatgrandchild0",
      "siblingsCount": 1,
      "title": "Child0_GrandChild0_GreatGrandChild0",
      "type": "tutorial",
      "uid": 366459535,
    },
    Object {
      "childUIDs": Array [],
      "depth": 1,
      "groupMarkerUID": -1893192488,
      "id": "Child0_GrandChild2",
      "index": 3,
      "parent": 365745369,
      "path": "/documentation/foo/child0/grandchild2",
      "siblingsCount": 4,
      "title": "Child0_GrandChild2",
      "type": "tutorial",
      "uid": -2044519617,
    },
    Object {
      "childUIDs": Array [
        1069111095,
      ],
      "depth": 0,
      "groupMarkerUID": -196255993,
      "id": "Child1",
      "index": 2,
      "parent": <root>,
      "path": "/documentation/foo/child1/",
      "siblingsCount": 3,
      "title": "Child1",
      "type": "tutorial",
      "uid": -1561138820,
    },
    Object {
      "beta": true,
      "childUIDs": Array [],
      "depth": 1,
      "id": "Child1_GrandChild0",
      "index": 0,
      "parent": -1561138820,
      "path": "/documentation/foo/child1/grandchild0",
      "siblingsCount": 1,
      "title": "Child1_GrandChild0",
      "type": "method",
      "uid": 1069111095,
    },
  ],
  "isFetching": false,
  "isFetchingAPIChanges": false,
  "references": undefined,
  "technology": Object {
    "children": Array [
      Object {
        "id": "Group Marker",
        "title": "Group Marker",
        "type": "groupMarker",
      },
      Object {
        "beta": true,
        "children": Array [
          Object {
            "id": "Group Marker, Child 0",
            "title": "Group Marker, Child 0",
            "type": "groupMarker",
          },
          Object {
            "beta": true,
            "id": "Child0_GrandChild0",
            "path": "/documentation/foo/child0/grandchild0",
            "title": "Child0_GrandChild0",
            "type": "tutorial",
          },
          Object {
            "children": Array [
              Object {
                "beta": true,
                "id": "Child0_GrandChild0_GreatGrandChild0",
                "path": "/documentation/foo/child0/grandchild0/greatgrandchild0",
                "title": "Child0_GrandChild0_GreatGrandChild0",
                "type": "tutorial",
              },
            ],
            "id": "Child0_GrandChild1",
            "path": "/documentation/foo/child0/grandchild1",
            "title": "Child0_GrandChild1",
            "type": "tutorial",
          },
          Object {
            "id": "Child0_GrandChild2",
            "path": "/documentation/foo/child0/grandchild2",
            "title": "Child0_GrandChild2",
            "type": "tutorial",
          },
        ],
        "id": "Child0",
        "path": "/documentation/foo/child0",
        "title": "Child0",
        "type": "article",
      },
      Object {
        "children": Array [
          Object {
            "beta": true,
            "id": "Child1_GrandChild0",
            "path": "/documentation/foo/child1/grandchild0",
            "title": "Child1_GrandChild0",
            "type": "method",
          },
        ],
        "id": "Child1",
        "path": "/documentation/foo/child1/",
        "title": "Child1",
        "type": "tutorial",
      },
    ],
    "path": "/documentation/foo",
  },
}
`;
